In [2]:
%matplotlib inline
import pandas as pd

In [2]:
from IPython.core.display import HTML
css = open('style-table.css').read() + open('style-notebook.css').read()
HTML('<style>{}</style>'.format(css))

In [80]:
titles = pd.read_csv('titles.csv')
titles.head()

title  year
0    Tasveer Mere Sanam  1996
1              Only You  1994
2  El pueblo del terror  1970
3               Machine  2007
4                  MARy  2008

### How many movies are listed in the titles dataframe?

In [15]:
titles.count()

title    244914
year     244914
dtype: int64

Movies are listed in the titles dataframe 244914

### What are the earliest two films listed in the titles dataframe?

In [21]:
titles.sort_values(by='year')[:2]

title  year
193222          Miss Jerry  1894
51388   The Startled Lover  1898

Miss Jerry	1894

The Startled Lover	1898

### How many movies have the title "Hamlet"?

In [33]:
s = titles[titles.title=='Hamlet']
s = titles[titles['title']=='Hamlet']

s.count()


title    20
year     20
dtype: int64

20 movies have the title 'Hamlet'

### How many movies are titled "North by Northwest"?

In [34]:
n = titles[titles.title=='North by Northwest']
n.count()

title    1
year     1
dtype: int64

1 movie has the title 'North by Northwest'

### When was the first movie titled "Hamlet" made?

In [38]:
s = titles[titles.title=='Hamlet']
s.sort_values(by='year')[:1]

title  year
52521  Hamlet  1910

### List all of the "Treasure Island" movies from earliest to most recent.

In [39]:
s = titles[titles.title=='Treasure Island']
s.sort_values(by='year')

title  year
55686   Treasure Island  1918
196514  Treasure Island  1920
141822  Treasure Island  1934
77354   Treasure Island  1950
65051   Treasure Island  1972
37238   Treasure Island  1973
173051  Treasure Island  1985
27076   Treasure Island  1999

### How many movies were made in the year 1950?

In [41]:
s = titles[titles.year==1950]
s.count()

# No.  of movies made in 1950 => 1113 

title    1113
year     1113
dtype: int64

### How many movies were made in the year 1960?

In [42]:
s = titles[titles.year==1960]
s.count()

# No.  of movies made in 1960 => 1549 

title    1549
year     1549
dtype: int64

### How many movies were made from 1950 through 1959?

In [73]:
s = titles[(titles.year>=1950) & (titles.year < 1960)]
s.count()

# 12394 movies


title    12934
year     12934
dtype: int64

### In what years has a movie titled "Batman" been released?

In [81]:
s = titles[titles.title=='Batman']
s['year']

# 1943 1989

39349     1943
240896    1989
Name: year, dtype: int64

In [3]:
cast = pd.read_csv('cast.csv')
cast.head()


title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

In [ ]:
### How many roles were there in the movie "Inception"?

In [109]:
s = cast[cast.title=='Inception']

# print(s.to_string())
s['type'].count()

# 78 roles

78

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [103]:
s = cast[cast.title=='Inception']
s['n'].isna().sum()

# 27 

27

### But how many roles in the movie "Inception" did receive an "n" value?

In [105]:
s = cast[cast.title=='Inception']
s['n'].notnull().sum()

# 51

51

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [116]:
s = cast[cast.title=='North by Northwest']

s = s.dropna(subset=['n']).sort_values(by='n')
print(s.to_string())

                      title  year                  name     type                    character     n
871426   North by Northwest  1959            Cary Grant    actor           Roger O. Thornhill   1.0
3518333  North by Northwest  1959       Eva Marie Saint  actress                  Eve Kendall   2.0
1457165  North by Northwest  1959           James Mason    actor              Phillip Vandamm   3.0
3157232  North by Northwest  1959   Jessie Royce Landis  actress              Clara Thornhill   4.0
356707   North by Northwest  1959        Leo G. Carroll    actor                The Professor   5.0
3048272  North by Northwest  1959  Josephine Hutchinson  actress                Mrs. Townsend   6.0
1695173  North by Northwest  1959           Philip Ober    actor              Lester Townsend   7.0
1275337  North by Northwest  1959         Martin Landau    actor                      Leonard   8.0
2447081  North by Northwest  1959         Adam Williams    actor                     Valerian   9.0


### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [119]:
s = cast[cast.title=='Sleuth']

s = s.sort_values(by='n')
print(s.to_string())

          title  year                 name     type                   character    n
326774   Sleuth  2007        Michael Caine    actor                      Andrew  1.0
1705394  Sleuth  1972     Laurence Olivier    actor                 Andrew Wyke  1.0
326773   Sleuth  1972        Michael Caine    actor                 Milo Tindle  2.0
1293634  Sleuth  2007             Jude Law    actor                        Milo  2.0
373876   Sleuth  1972       Alec Cawthorne    actor           Inspector Doppler  3.0
1805907  Sleuth  2007        Harold Pinter    actor                 Man on T.V.  3.0
1465483  Sleuth  1972   John (II) Matthews    actor  Detective Sergeant Tarrant  4.0
2724553  Sleuth  1972   Eve (III) Channing  actress             Marguerite Wyke  5.0
1448690  Sleuth  1972         Teddy Martin    actor      Police Constable Higgs  6.0
260242   Sleuth  2007      Kenneth Branagh    actor           Other Man on T.V.  NaN
373877   Sleuth  2007  Alec (II) Cawthorne    actor           Ins

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [120]:
s = cast[(cast.title=='Sleuth') & (cast.year== 2007)]

s = s.sort_values(by='n')
print(s.to_string())

          title  year                 name     type          character    n
326774   Sleuth  2007        Michael Caine    actor             Andrew  1.0
1293634  Sleuth  2007             Jude Law    actor               Milo  2.0
1805907  Sleuth  2007        Harold Pinter    actor        Man on T.V.  3.0
260242   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V.  NaN
373877   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler  NaN
2724552  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke  NaN
3370424  Sleuth  2007    Carmel O'Sullivan  actress             Maggie  NaN


### How many roles were credited in the silent 1921 version of Hamlet?

In [130]:
s = cast[(cast.title=='Hamlet') & (cast.year== 1921)]
print(s.to_string())
s = s.dropna(subset=['n']).sort_values(by='n')

s.count()

# 9 roles

          title  year                    name     type         character    n
7789     Hamlet  1921        Fritz Achterberg    actor       Fortinbras,  9.0
458904   Hamlet  1921            Paul Conradi    actor      K?nig Hamlet  2.0
555034   Hamlet  1921        Anton De Verdier    actor           Laertes  7.0
1135279  Hamlet  1921         Hans Junkermann    actor          Polonius  6.0
2185558  Hamlet  1921            Heinz Stieda    actor           Horatio  5.0
2375993  Hamlet  1921  Eduard von Winterstein    actor          Claudius  4.0
2664423  Hamlet  1921         Mathilde Brandt  actress  K?nigin Gertrude  3.0
3062023  Hamlet  1921          Lilly Jacobson  actress           Ophelia  8.0
3355176  Hamlet  1921            Asta Nielsen  actress            Hamlet  1.0


title        9
year         9
name         9
type         9
character    9
n            9
dtype: int64

### How many roles were credited in Branagh’s 1996 Hamlet?

In [134]:
s = cast[(cast.title=="Hamlet") & (cast.year== 1996)]
s = s.dropna(subset=['n']).sort_values(by='n')
# print(s.to_string())

s['n'].count()

# 46 roles

46

### How many "Hamlet" roles have been listed in all film credits through history?

In [135]:
s = cast[cast.title=="Hamlet"]
s = s.dropna(subset=['n']).sort_values(by='n')
s['n'].count()

# 249

249

### How many people have played an "Ophelia"?

In [138]:
s = cast[cast.character=="Ophelia"]
s.character.count()

# 117

117

### How many people have played a role called "The Dude"?

In [139]:
s = cast[cast.character=="The Dude"]
s.character.count()

# 19

19

### How many people have played a role called "The Stranger"?

In [140]:
s = cast[cast.character=="The Stranger"]
s.character.count()

# 212

212

### How many roles has Sidney Poitier played throughout his career?

In [143]:
s = cast[cast.name=="Sidney Poitier"]
# print(s.to_string())
s.name.count()

# 43

43

### How many roles has Judi Dench played?

In [144]:
s = cast[cast.name=="Judi Dench"]
# print(s.to_string())
s.name.count()

# 55

55

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [154]:
s = cast[(cast.name=="Cary Grant") & (cast.year >= 1940) & (cast.year < 1950)]
s = s[s.n == 2.0]

print(s.to_string())

# 2

                   title  year        name   type    character    n
871423  My Favorite Wife  1940  Cary Grant  actor   Nick Arden  2.0
871433    Penny Serenade  1941  Cary Grant  actor  Roger Adams  2.0


### List the leading roles that Cary Grant played in the 1940s in order by year.

In [161]:
s = cast[(cast.name=="Cary Grant") & (cast.year >= 1940) & (cast.year < 1950)]
s = s[s.n == 1.0]
s = s.sort_values(by='year')
s[['name', 'year', 'title']]


name  year                                 title
871448  Cary Grant  1940               The Howards of Virginia
871407  Cary Grant  1940                       His Girl Friday
871450  Cary Grant  1940                The Philadelphia Story
871438  Cary Grant  1941                             Suspicion
871452  Cary Grant  1942                  The Talk of the Town
871429  Cary Grant  1942                 Once Upon a Honeymoon
871398  Cary Grant  1943                     Destination Tokyo
871422  Cary Grant  1943                             Mr. Lucky
871430  Cary Grant  1944                      Once Upon a Time
871390  Cary Grant  1944                  Arsenic and Old Lace
871425  Cary Grant  1944             None But the Lonely Heart
871424  Cary Grant  1946                         Night and Day
871427  Cary Grant  1946                             Notorious
871444  Cary Grant  1947      The Bachelor and the Bobby-Soxer
871445  Cary Grant  1947                     The Bishop's Wife
871421  Cary Grant  1948  Mr. Blandings Builds His Dream House
871402  Cary Grant  1948          Every Girl Should Be Married
871411  Cary Grant  1949                I Was a Male War Bride

### How many roles were available for actors in the 1950s?

In [165]:
s = cast[(cast.type=="actor") & (cast.year >= 1950) & (cast.year < 1960)]
s['type'].count()

# 157738

157738

### How many roles were available for actresses in the 1950s?

In [166]:
s = cast[(cast.type=="actress") & (cast.year >= 1950) & (cast.year < 1960)]
s['type'].count()

# 57733


57733

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [6]:
s = cast[cast.year <= 1980]
s = s.dropna(subset=['n'])

s = s[s.n == 1.0]

print(s.to_string())

                                      title  year              name     type                 character    n
154605       The Fairylogue and Radio-Plays  1908     L. Frank Baum    actor      The Wizard of Oz Man  1.0
794541   The Joe Gans-Battling Nelson Fight  1906          Joe Gans    actor                   Himself  1.0
818490                   Robbery Under Arms  1907        Jim Gerald    actor                  Warrigal  1.0
952438                    The Life of Moses  1909      Pat Hartigan    actor                     Moses  1.0
1272777                       Lika mot lika  1906      Knut Lambert    actor                 Coq-H?ron  1.0
2371817                       Pyotr Velikiy  1910      Pyotr Voinov    actor  Pyotr I, Peter the Great  1.0
2841510             La petite Blanche-Neige  1910  H?l?ne Du Montel  actress             Blanche-Neige  1.0
2860093                           Jane Eyre  1910       Marie Eline  actress      Jane Eyre as a Child  1.0
3183324                Lucre

### How many non-leading roles were available through from the beginning of film history through 1980?

In [5]:
s = cast[cast.year <= 1980]
s = s.dropna(subset=['n'])
s = s[s.n != 1.0]

print(s.to_string())

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [10]:
s = cast[cast.year <= 1980]
s['n'].isna().count()

# print(s.to_string())

1182807